main.cpp

```diff
@@ -176,108 +121,394 @@ int main()
 
  Wait for my code review.
  */
-
 #include <iostream>
+
+// Forward declare new types
+struct FloatType;
+struct DoubleType;
+struct IntType;
+
 struct FloatType
 {
-    float add(float lhs = 0.f, float rhs = 0.f );
-    float subtract( float lhs = 0.f, float rhs = 0.f );
-    float multiply( float lhs = 0.f, float rhs = 0.f );
-    float divide( float lhs = 0.f, float rhs = 0.f );
+    float* value = nullptr;
+    
+    FloatType(float v)
+    {
+        value = new float(v);
+    }
+    ~FloatType()
+    {
+        delete value;
+    }
+    FloatType& add( float rhs = 0.f );
+    FloatType& subtract( float rhs = 0.f );
+    FloatType& multiply( float rhs = 0.f );
+    FloatType& divide( float rhs = 0.f );
+    FloatType& add( const FloatType& rhs );
+    FloatType& subtract( const FloatType& rhs );
+    FloatType& multiply( const FloatType& rhs );
+    FloatType& divide( const FloatType& rhs );
+    FloatType& add( const DoubleType& rhs );
+    FloatType& subtract( const DoubleType& rhs );
+    FloatType& multiply( const DoubleType& rhs );
+    FloatType& divide( const DoubleType& rhs );
+    FloatType& add( const IntType& rhs );
+    FloatType& subtract( const IntType& rhs );
+    FloatType& multiply( const IntType& rhs );
+    FloatType& divide( const IntType& rhs );
+};
+
+struct DoubleType
+{
+    double* value = nullptr;
+    DoubleType(double v)
+    {
+        value = new double(v);
+    }
```

Comment:
When you have to choose where to initialize your member variables and your choices are:
- In-class
- Constructor Member Initializer List
- Inside the Constructor's body

Try to use In-class initialization, followed by the Constructor Member Initializer List.
The reason is that the member variables need to be instantiated/initialized when the class is created.
When you assign to them in the constructor body, you are potentially re-initializing them, which is a waste of resources.
Take a look at the output of this program: https://godbolt.org/z/q83YM63ze
Take note of the second instance being created and destroyed due to that constructor body assignment.
- [ ] revise, use the member initializer list to initialize your member variables